### I- Scores avant ajout du SQL sur le RAG

#### 1 - Scores détaillés par ligne 

In [13]:
import pandas as pd

In [14]:
score_1 = pd.read_csv("ragas_results.csv")
score_1

,question,answer,contexts,ground_truth,faithfulness,answer_relevancy,context_precision,context_recall
0,Quel joueur a le meilleur pourcentage à 3 poin...,"D'après les données disponibles, Shai Gilgeous...",['Analyse des performances des joueurs en tena...,Alondes Williams (DET) affiche un pourcentage ...,0.5,0.709660,1.0,0.000000
1,Quel joueur a le meilleur % de réussite aux la...,"D'après les données disponibles, Shai Gilgeous...",['Comparaison des performances des joueurs en ...,Sam Hauser (BOS) affiche un taux de réussite d...,1.0,0.901402,1.0,0.333333
2,Quel joueur a le plus de points combinés (poin...,"D'après les données disponibles, Nikola Jokić ...",['Agrégation de plusieurs statistiques pour év...,Nikola Jokić domine nettement avec 3675 unités...,NaN,0.000000,1.0,0.333333
3,Quel joueur a la meilleure efficacité offensiv...,"D'après les données disponibles, Joel Embiid (...",['Analyse des performances ciblées contre des ...,Giannis Antetokounmpo (MIL) est le joueur le p...,1.0,0.754911,1.0,0.000000
4,Quel joueur a amélioré son pourcentage à 2 poi...,"D'après les données disponibles, Shai Gilgeous...","[""Analyse de l'évolution des statistiques d'un...",Alondes Williams (DET) présente un taux de réu...,1.0,0.000000,0.0,0.000000
5,Quelle est la durée légale d’un CDD en France ?,"En tant qu'analyste NBA, je ne suis pas compét...",['La durée maximale d’un CDD en France est gén...,La durée maximale d’un CDD est en général de 1...,NaN,0.000000,1.0,1.000000
6,Comment obtenir un passeport biométrique ?,"En tant qu'analyste NBA, je me concentre sur l...","[""Un passeport biométrique s'obtient en mairie...",L’obtention d’un passeport biométrique nécessi...,1.0,0.860621,1.0,1.000000


#### 2- Scores globaux

In [15]:
score_1_1 = pd.read_csv("ragas_summary.csv")
score_1_1

,Metric,Average,Std
0,faithfulness,0.900000,0.223607
1,answer_relevancy,0.460942,0.435781
2,context_precision,0.857143,0.377964
3,context_recall,0.380952,0.448395


### II- Scores avant ajout du SQL sur le RAG

#### 1- Scores détaillés

In [16]:
score_2 = pd.read_csv("ragas_results_2.csv")
score_2

,question,answer,contexts,ground_truth,faithfulness,answer_relevancy,context_precision,context_recall
0,Quel joueur a le meilleur pourcentage à 3 poin...,Alondes Williams (DET) affiche un incroyable 1...,['Analyse des performances des joueurs en tena...,Alondes Williams (DET) affiche un pourcentage ...,1.0,0.765055,0.0,0.000000
1,Quel joueur a le meilleur % de réussite aux la...,Sam Hauser (BOS) affiche un taux de réussite p...,['Comparaison des performances des joueurs en ...,Sam Hauser (BOS) affiche un taux de réussite d...,1.0,0.779099,1.0,0.333333
2,Quel joueur a le plus de points combinés (poin...,Nikola Jokić domine largement avec 3675 points...,['Agrégation de plusieurs statistiques pour év...,Nikola Jokić domine nettement avec 3675 unités...,1.0,0.779061,1.0,0.333333
3,Quel joueur a la meilleure efficacité offensiv...,Giannis Antetokounmpo (MIL) est le joueur le p...,['Analyse des performances ciblées contre des ...,Giannis Antetokounmpo (MIL) est le joueur le p...,1.0,0.957428,1.0,0.500000
4,Quel joueur a amélioré son pourcentage à 2 poi...,Alondes Williams (DET) affiche un pourcentage ...,"[""Analyse de l'évolution des statistiques d'un...",Alondes Williams (DET) présente un taux de réu...,1.0,0.758162,1.0,0.000000
5,Quelle est la durée légale d’un CDD en France ?,Ce domaine ne relève pas de mes compétences. V...,['La durée maximale d’un CDD en France est gén...,La durée maximale d’un CDD est en général de 1...,NaN,0.000000,1.0,1.000000
6,Comment obtenir un passeport biométrique ?,Ce domaine ne relève pas de mes compétences. V...,"[""Un passeport biométrique s'obtient en mairie...",L’obtention d’un passeport biométrique nécessi...,NaN,0.000000,1.0,1.000000


#### 2- Scores globaux

In [18]:
score_2_1 = pd.read_csv("ragas_summary_2.csv")
score_2_1

,Metric,Average,Std
0,faithfulness,1.000000,0.000000
1,answer_relevancy,0.576972,0.400092
2,context_precision,0.857143,0.377964
3,context_recall,0.452381,0.416270


### III- Interprétations générales

**✅ Comparaison des performances RAG — Avant vs Après SQL**

1. Faithfulness (fidélité factuelle)

    - Score avant SQL : 0.90
    - Score après SQL : 1.00

    **Analyse**

    L’ajout de la base SQL fait passer la fidélité à 100%, ce qui signifie que :

    les réponses du modèle sont désormais toujours factuellement alignées avec la vérité terrain (ground truth),

    le système évite mieux les hallucinations,

    la requête SQL apporte une source structurée et fiable, ce qui rend les réponses beaucoup plus exactes.

    C’est la plus forte amélioration observée.

    **Pistes d’amélioration**

    Même si le score est maximal, il reste possible de sécuriser durablement cette performance :

    Renforcer le prompt système pour interdire explicitement toute information hors contexte :

    “Si l’information n’est pas présente dans le contexte ou les résultats SQL, réponds explicitement que l’information est inconnue.”

    Forcer la citation implicite des sources (texte ou SQL) dans la réponse afin d’obliger le modèle à s’appuyer sur des faits vérifiables.

    Ajouter des exemples question–réponse contraignants dans le prompt (few-shot), montrant des cas où le modèle doit refuser de répondre en l’absence de données suffisantes.

    ***Ces actions permettent de maintenir un score élevé même lorsque les questions deviennent plus complexes ou ambiguës.***

2. Answer Relevancy (pertinence de la réponse)

    - Score avant SQL : 0.4609
    - Score après SQL : 0.5769

    **Analyse**

    Une amélioration d’environ +25%, ce qui montre que :

    les réponses sont maintenant plus en phase avec ce que l’utilisateur demande,

    le modèle comprend mieux la question car il peut s'appuyer sur des données SQL structurées,

    la présence de vraies statistiques issues de la base SQL donne des réponses plus ciblées.

    ➡️ Le système devient plus utile, pas seulement plus exact.

    **Pistes d’amélioration**

    Ce score reste perfectible et peut être amélioré par :

    Un prompt plus orienté “réponse directe”, par exemple :

    demander explicitement une réponse concise,

    structurer la réponse (liste, phrases courtes, chiffres clés).

    L’ajout d’exemples question → réponse attendue (few-shot prompting) afin d’illustrer clairement le niveau de détail attendu.

    Un pré-traitement des questions utilisateur (reformulation automatique) pour réduire les ambiguïtés avant le passage au retriever et au SQL.

    Un meilleur mapping NL → SQL, afin de s’assurer que la requête SQL répond exactement à l’intention de la question.

    ***Ces améliorations visent à réduire l’écart entre ce que l’utilisateur attend et ce que le modèle produit.***

3. Context Precision

    - Score avant SQL : 0.8571
    - Score après SQL : 0.8571

    **Analyse**

    Le score ne change pas.

    C’est logique :

    la logique de sélection des chunks RAG initiaux n’a pas été modifiée,

    l’ajout de SQL a simplement enrichi les sources sans influencer la précision du retriever vectoriel.

    ➡️ Stabilité normale, indiquant que le retriever était déjà performant.

    **Pistes d’amélioration**

    Même avec un bon score, des optimisations sont possibles :

    Affiner la taille des chunks pour éviter les passages trop longs contenant des informations inutiles.

    Réduire légèrement le top_k si certains chunks apportent peu de valeur informative.

    Améliorer les embeddings (modèle plus récent ou embeddings spécialisés domaine NBA).

    ***L’objectif est de réduire encore le bruit tout en conservant l’information utile.***

4. Context Recall

    - Score avant SQL : 0.3809
    - Score après SQL : 0.4523

    **Analyse**

    Une amélioration de +18%, signe que :

    les réponses utilisent désormais davantage de contextes pertinents,

    les requêtes SQL comblent certaines lacunes laissées par la recherche vectorielle,

    le modèle peut exploiter un ensemble d’informations plus complet.

    ➡️ Le système retrouve plus d’informations pertinentes pour répondre correctement.

    **Pistes d’amélioration**

    Le recall restant modéré, plusieurs leviers peuvent être actionnés :

    Augmenter légèrement le top_k du retriever pour élargir la couverture informationnelle.

    Améliorer la segmentation documentaire (chunks mieux alignés avec les thèmes des questions).

    Enrichir la base documentaire avec des sources plus complètes ou mieux structurées.

    ***Ces actions permettent de ne pas passer à côté d’informations essentielles.***

**Interprétation générale**

L’ajout de la base SQL a :

Éliminé les hallucinations (faithfulness = 1.00),

Rendu les réponses plus pertinentes (answer relevancy ↑),

Amélioré l’exploitation du contexte (context recall ↑),

Sans dégrader la précision du contexte (context precision stable).

***Le système RAG est désormais plus fiable, plus pertinent et plus robuste, tout en offrant des pistes claires d’amélioration pour les prochaines itérations.***

### IV - Validation des étapes avec Pydantic

| **Étape du pipeline**                           | **Éléments exacts du code impliquant Pydantic / Logfire**                        | **Rôle**                                                                            |
| ----------------------------------------------- | -------------------------------------------------------------------------------- | --------------------------------------------------------------------------------------------------------------------------- |
| **1. Lecture des fichiers (.pdf / .txt / .md)** | `DocumentIn` (Pydantic BaseModel) + `field_validator('text')`                    | ✔ Valide que `doc_id` et `text` existent<br>✔ Empêche les documents vides<br>✔ Nettoie automatiquement le texte (`strip()`) |
| **2. Logfire (simulé) — début du chargement**   | `_logfire_span("chargement_fichiers")`                                           | ✔ Trace l'entrée dans l'étape de chargement<br>✔ Ajoute des métadonnées dans les logs                                       |
| **3. Extraction PDF**                           | `_logfire_span("extraction_texte_pdf")`                                          | ✔ Trace l’extraction PDF, erreurs, succès, nb pages                                                                         |
| **4. Création de DocumentIn**                   | `document = DocumentIn(...)`                                                     | ✔ Structure les données du document<br>✔ Valide métadonnées (file_type, char_count...)                                      |
| **5. Logfire — Validation des fichiers**        | `_logfire_span("pdf_charge")` / `_logfire_span("texte_charge")`                  | ✔ Suivi détaillé (nombre de caractères, taille fichier)                                                                     |
| **6. Nettoyage des documents**                  | `clean_documents()` + Pydantic (`doc.text = cleaned_text`)                       | ✔ Pydantic re-valide les objets lors de l'affectation<br>✔ Filtre les documents trop courts                                 |
| **7. Logfire — Nettoyage**                      | `_logfire_span("nettoyage_documents")` + `_logfire_span("document_conserve")`    | ✔ Trace combien de documents nettoyés/ignorés                                                                               |
| **8. Chunking**                                 | Création de `Chunk()` (Pydantic BaseModel)                                       | ✔ Empêche chunks vides<br>✔ Valide indices (start_index/end_index ≥ 0)<br>✔ Propage les métadonnées du document             |
| **9. Logfire — Découpage**                      | `_logfire_span("decoupage_document")` + `_logfire_span("document_decoupe")`      | ✔ Trace nb de chunks créés par document                                                                                     |
| **10. Embeddings**                              | Création de `EmbeddingVec()` (Pydantic BaseModel)                                | ✔ Valide que chaque vecteur existe<br>✔ Empêche vecteurs vides<br>✔ Assure que chunk_id existe                              |
| **11. Logfire — Embeddings**                    | `_logfire_span("generation_embeddings")` + `_logfire_span("embeddings_generes")` | ✔ Trace nombre de vecteurs et dimension<br>✔ Trace erreurs éventuelles                                                      |
| **12. Pipeline complet**                        | `_logfire_span("pipeline_complet")` et spans sur chaque étape                    | ✔ Trace toutes les étapes du pipeline du début à la fin<br>✔ Permet un suivi détaillé et un debug complet                   |


| **Étape du pipeline**                           | **Rôle**                                                                                                                   |  
|-------------------------------------------------| -------------------------------------------------------------------------------------------------------------------------- | 
| **1. Lecture des fichiers (.pdf / .txt / .md)** | ✔ Valide que `doc_id` et `text` existent<br>✔ Empêche les documents vides<br>✔ Nettoie automatiquement le texte (`strip()`) | 
| **2. Logfire (simulé) — début du chargement**   |  ✔ Trace l'entrée dans l'étape de chargement<br>✔ Ajoute des métadonnées dans les logs                                       |
| **3. Extraction PDF**                           |  ✔ Trace l’extraction PDF, erreurs, succès, nb pages                                                                         |
| **4. Création de DocumentIn**                   |  ✔ Structure les données du document<br>✔ Valide métadonnées (file_type, char_count...)                                      |
| **5. Logfire — Validation des fichiers**        |  ✔ Suivi détaillé (nombre de caractères, taille fichier)                                                                     |
| **6. Nettoyage des documents**                  |  ✔ Pydantic re-valide les objets lors de l'affectation<br>✔ Filtre les documents trop courts                                 |
| **7. Logfire — Nettoyage**                      |  ✔ Trace combien de documents nettoyés/ignorés                                                                               |
| **8. Chunking**                                 |  ✔ Empêche chunks vides<br>✔ Valide indices (start_index/end_index ≥ 0)<br>✔ Propage les métadonnées du document             |
| **9. Logfire — Découpage**                      |  ✔ Trace nb de chunks créés par document                                                                                     |
| **10. Embeddings**                              | ✔ Valide que chaque vecteur existe<br>✔ Empêche vecteurs vides<br>✔ Assure que chunk_id existe                              |
| **11. Logfire — Embeddings**                    |  ✔ Trace nombre de vecteurs et dimension<br>✔ Trace erreurs éventuelles                                                      |
| **12. Pipeline complet**                        |  ✔ Trace toutes les étapes du pipeline du début à la fin<br>✔ Permet un suivi détaillé et un debug complet                   |


### V- Structure commplète du RAG

┌────────────────────────────────────────────┐
│              SOURCES DE DONNÉES            │
└────────────────────────────────────────────┘
              │
   ┌──────────▼──────────┐
   │  Documents textuels │      ┌───────────────────── ─┐
   │  (PDF, Excel, CSV)  │      │  Base SQL (Postgres)  │
   │  Données NBA        │      │  Stats, joueurs,      │
   └──────────┬──────────┘      │  équipes, matchs      │
              │                 └──────────┬─────────── ┘
              │                            │
┌─────────────▼──────────── ─┐   ┌──────────▼───────────┐
│       Chunking             │   │    SQL Tool / NL→SQL │
│  (découpage des textes)    │   │  (LangChain / Prompt)│
└─────────────┬──────────── ─┘   └──────────┬───────────┘
              │                             │
┌─────────────▼───────────── ┐   ┌──────────▼───────────┐
│     Embeddings (Mistral)   │   │   Requêtes SQL       │
│  text → vecteurs numériques│   │   Résultats fiables  │
└─────────────┬───────────── ┘   └──────────┬───────────┘
              │                             │
┌─────────────▼──────────── ─┐   ┌──────────▼───────────┐
│   Vector Store (FAISS)     │◄──┤   Enrichissement     │
│   Index vectoriel          │   │   des contextes      │
│   Similarité sémantique    │   │   (SQL → texte)      │
└─────────────┬───────────── ┘   └──────────────────────┘
              │
┌─────────────▼─────────────┐
│        Retriever          │
│  Top-k chunks pertinents  │
│  + données SQL enrichies  │
└─────────────┬─────────────┘
              │
┌─────────────▼─────────────┐
│      Prompt Builder       │
│  Question utilisateur     │
│  + contexte vectoriel     │
│  + résultats SQL          │
└─────────────┬─────────────┘  
              │
┌─────────────▼──────────── ─┐
│        LLM (Mistral)       │
│  Raisonnement contrôlé     │
│  Réponse fidèle au contexte│
└─────────────┬──────────── ─┘
              │
┌─────────────▼─────────────┐
│       Réponse finale      │
│  Précise, pertinente,     │
│  sans hallucination       │
└───────────────────────────┘
